Add lazy initialization and isLookupEnabled

jvican opened the pull request:
- Now, generated picklers/unpicklers are initialised only if there is no
  pickler/unpickler already registered for a given type. If that's the
  case, the registered on is reused.
- There's a special handling for the generation of both Pickler and
  Unpickler. We have to make sure that both pickler/unpickler exist and
  that they are the same one, since we have to return a class that
  matches the return type (`Pickler[T]` with `Unpickler[T]`).
- As a side note, it's necessary to cast `Unpickler`s and
  `Pickler with Unpickler` to `Generated`, otherwise the implicit search
  fails.
- Add a field in the `PicklerRegistry` that check if the lookup is
  enabled for every kind of registr.

jvican committed:
Allow `isLookupEnabled` in the `PicklerRegistry`

* Add a field in the `PicklerRegistry` that check if the lookup is
  enabled for every kind of registr.

jvican committed:
Add lazy initialization of generated (un)picklers

* Now, generated picklers/unpicklers are initialised only if there is no
  pickler/unpickler already registered for a given type. If that's the
  case, the registered on is reused.

* There's a special handling for the generation of both Pickler and
  Unpickler. We have to make sure that both pickler/unpickler exist and
  that they are the same one, since we have to return a class that
  matches the return type (`Pickler[T]` with `Unpickler[T]`).

* As a side note, it's necessary to cast `Unpickler`s and
  `Pickler with Unpickler` to `Generated`, otherwise the implicit search
  fails.

jvican commented:
It depends on #408 and therefore contains its commit messages as well. When the parent is merged, I will rebase. Please, check only the other two last commits.

jvican committed:
Don't check existence in `AutoRegister`

* Because of lazy initialization, checks are not necessary anymore.

jvican commented:
Failed tests are somehow related to `shareNothing`. I don't really know what's going on here, since @jsuereth has pointed out in the past some problems with `shareNothing` and `refPicklers` in the new release.

The tests are expecting memory-related exceptions that don't get to be thrown but correctly (un)pickled. The following comment explains that there were some errors thrown by Jenkins: https://github.com/scala/pickling/blob/0.11.x/core/src/test/scala/pickling/run/share-json.scala#L64.

Interestingly, all the tests pass if run using Intellij.

jvican commented:
No, actually these tests should fail, there's a circular dependency and `shareNothing` is in scope. So I guess that this is uncovering a new issue with it, isn't it?

jvican commented:
My guess of what's happening: Intellij and sbt runs tests differently. Somehow, with sbt, the tests share some global state. Therefore, it is first executed and registered the (un)pickler for `C` which was created with `shareEverything` in scope. Later, when the (un)pickler for `C` that uses `shareNothing` is invoked, it reuses the registered one.

**EDIT**: My guess has been confirmed. Working on a fix for this.

jsuereth commented on core/src/main/scala/scala/pickling/internal/DefaultPicklerRegistry.scala:
```diff
@@ -52,7 +52,15 @@ final class DefaultPicklerRegistry(generator: RuntimePicklerGenerator) extends P
 
-  /** Checks the existince of an unpickler. */
+  /** Checks the existence of a pickler ignoring the registered generators. */
+  override def lookupExistingPickler(key: String): Option[Pickler[_]] =
```
Note: Generators are for type constructors and SHOULD be safe to call.  They do not involve runtime reflection.

E.g.  `List[_]` is impossible to register a pickler for, but you _can_ register a pickler that handles it _if_ it can lookup a pickler for element types.

Note:  I may be adding statically generated "generator" functions for types like `List[_]` where you leave some of the type constructor args "empty" and we get them passed in at runtime and do a runtime lookup (but not runtime reflection).

In any case, you shoudl be using lookupPickler/lookupUnpickler.

jvican replied:
Yes, I know they do not involve runtime reflection. But we want that a concrete pickler has preference over the generator. That is to say that `Pickler[List[String]]` should be returned and registered instead of returning a generator `Pickler[List[_]]`. That's why these methods exist: we want to check when we initialize concrete (un)picklers that there is no _other_ concrete (un)pickler for that type. This was done with #410 on mind, thus those tests will pass.

jsuereth commented on core/src/main/scala/scala/pickling/generator/sourcegen.scala:
```diff
@@ -384,63 +384,114 @@ private[pickling]  trait SourceGenerator extends Macro with tags.FastTypeTagMacr
   }
 
   val picklingPath = q"_root_.scala.pickling"
+  val picklersRegistry = q"$picklingPath.internal.currentRuntime.picklers"
```
nice! this makes things a lot cleaner, I'll have to mimick this elsewhere.

jsuereth commented on core/src/main/scala/scala/pickling/generator/sourcegen.scala:
```diff
@@ -384,63 +384,114 @@ private[pickling]  trait SourceGenerator extends Macro with tags.FastTypeTagMacr
+    q"""
+      if($picklersRegistry.isLookupEnabled) {
+        ($lookup1($key), $lookup2($key)) match {
+          case (Some(p), Some(u)) if p == u => p.asInstanceOf[$tpe]
```
maybe document that `tpe` is the type of the _pickler_ not the type being pickled.

jvican replied:
Will do :+1:

jsuereth commented:
NICE!  Generally LGTM, but one caveat:  Tests are failing.

This is another instance where because we have a global runtime registry of picklers, you'll need to "swap" or "clear" the register in the test to make sure we can generate a _new_ pickler with sharing disabled.

It _may_ make sense to add a "clear" method to to the pickler registry just for testing.  You could probably make it `private[pickling]` so that only our tests can call it.  I'm not sure it makes sense in other contexts....

Option #2 is to implement a thread-local pickler registry and swap threads.   That seems pretty crazy and not a good idea....

jvican commented:
I think there is even a better solution: to keep track of what picklers/unpicklers are generated with `SharedNothing` and check in the lazy initialization if there's a mismatch between the Shared options of the pickler that we want to initialize and the one that is registered.

I'll add the `clear` method but I believe that my solution is more focused on the long term, because other similar issues could appear in production code (I've thought of other use cases in which this could fail).

WDYT? BTW, I'm working on this right now, i'll have it in 10 min or so.

jsuereth commented:
@jvican Well, considering I'm thinking of deprecating the way we handle sharing (because it's subtly broken for most statically generated picklers), I'm not sure we want to expose more sharing-related code :)

That said, tracking in the pickler whether or not it supports sharing is probably best.   That said, if you have one pickler with sharing and another without in the same runtime, isn't that a logic bug in your program?

jvican commented:
It depends... going again to the analogy of client-server, wouldn't be good to support the use case of having a client that fully knows the message that is sending (and therefore enables shareNothing) and a server (who doesn't know anything about the message received, but knows that it could be either a tree or graph, and therefore enables shareEverything)?

Also, it's not necessary to keep track of that in the (un)picklers, we can do it just in the `PicklerRegistry` by asking for an implicit of `Share`. I think it's pretty neat and it's not a lot of code, so we can remove it whenever you tackle the refactoring of the sharing :smile:

jsuereth commented:
@jvican So the client/server analogy falls apart here. The issue only arises same JVM.  The reasons tests fail is those tests do teh following (in sequence):
1. Register sharing picklers
2. pickle and validate
3. Register non-sharing picklers (oops, we can't do that now)
4. pickle and validate

So for testing reasons, being able to "clean" the registered picklers between tests may be useful.  Eitehr that _Or_ when we test non-sharing we create totally new types/classees.   That's probably the easiest (and least hacky) solution :)

jvican commented:
Your suggestion of creating totally new types for non-sharing (un)picklers is great. I think that it's very similar to my current solution (which I'm about to submit) but I don't embed that "sharing" information in the type, just in a dictionary.

Tests are passing now in my machine. Otherwise, if you prefer to strictly follow the 'embed-in-the-types' strategy I can give it a try!

jsuereth commented:
let me see what you have, if it's almost done, then I'll comment :)  Hopefully it's good to go.

jsuereth commented on core/src/main/scala/scala/pickling/internal/DefaultPicklerRegistry.scala:
```diff
@@ -14,10 +14,14 @@ final class DefaultPicklerRegistry(generator: RuntimePicklerGenerator) extends P
+  private val unpicklerMap: Mmap[String, Unpickler[_]] = new TrieMap[String, Unpickler[_]]
+  private val unpicklerGenMap: Mmap[String, UnpicklerGenerator] = new TrieMap[String, UnpicklerGenerator]
+
+  private val shareNothingPicklers: Mset[String] = Mset.empty[String]
```
Given that most users will be picking ONE share strategy for their entire protocol, I do think this is a pretty big overhead/burden on the API/users for such a small use case (specifically the test).

jsuereth commented:
@jvican SO, I think pickling strategies are already complicated enough.  I think maybe we just fix the tests so the "share nothing" types are not the same as the 'share' tests...

jvican commented:
I agree, working on the `clear` solution!

jvican committed:
Fix clash between picklers with different sharing

* Different generated picklers with opposite sharing strategies are
  created and only the first one is registered. The second one never
  gets initiliazed and reuses the first one.

* Remove `freshName` since scala `2.10.0` doesn't support.

jvican commented:
Tests are passing with the `clear` trick.

jsuereth commented on core/src/main/scala/scala/pickling/spi/PicklerRegistry.scala:
```diff
@@ -98,5 +98,7 @@ trait PicklerRegistry {
   def registerPicklerUnpicklerGenerator[T](typeConstructorKey: String, generator: FastTypeTag[_] => (Pickler[T] with Unpickler[T])): Unit
-  // TODO - Some kind of clean or inspect what we have?
+
+  private[pickling] def clearRegisteredPicklerUnpicklerFor[T: FastTypeTag]: Unit
```
nit: Can you add some quick docs explaining this?  We understand, but someone new to the codebase will be curious.

jvican replied:
Totally forgot :smile:

jsuereth commented:
One minor nit, fix that and feel free to merge!  looking great.

jvican commented:
Tests passed. Merging this!
